dh/sidh/internal/templates/curve_test.gotemp

```diff
@@ -98,3 +102,207 @@ func BenchmarkThreePointLadder(b *testing.B) {
+func InvalidPKNoneLinear(scalar []uint8, pub3Pt *[3]Fp2) {
+	
+	var P, Q ProjectivePoint
+	
+	rand.Seed(time.Now().UnixNano())
+    random_index := rand.Intn(int(params.B.SecretByteLen-1)*8)
+	
+	// Set P as a point of order 3^e3
+    P = ProjectivePoint{X: params.B.AffineP, Z: params.OneFp2}
+
+	// Set Q = [k]P, where k = scalar[:random_index] 
+	Q = montgomeryLadder(&params.InitCurve, &P, scalar, uint(random_index))
+
+	// invQz = 1/Q.Z
+	var invQz Fp2
+	invQz = Q.Z
+	inv(&invQz, &invQz)
+
+	mul(&pub3Pt[0], &P.X, &P.Z)
+	mul(&pub3Pt[1], &Q.X, &invQz)
+	
+
+}
+
+// Construct Invalid public key tuple (P,Q) such that Q = [k]P + T, where k is random and T is the point of order 2. 
+// Simulate HB and section 3.1.2 of paper https://eprint.iacr.org/2022/054.pdf
+// We only construct point P and Q because in the attacks the third point is P-Q by construction 
+// and the countermeasure does not test it
+// Without loss of generality, we assume the curve is the starting curve 
+func InvalidPKT(scalar []uint8, pub3Pt *[3]Fp2) {
+	var P, Q ProjectivePoint
+	
+	rand.Seed(time.Now().UnixNano())
+    random_index := rand.Intn(int(params.B.SecretByteLen-1)*8)
+	
+	// Set P as a point of order 3^e3
+    P = ProjectivePoint{X: params.B.AffineP, Z: params.OneFp2}
+
+	// Set Q = [k]P, where k = scalar[:random_index] 
+	Q = montgomeryLadder(&params.InitCurve, &P, scalar, uint(random_index))
+
+	// Q = [k]P + T
+	tauT(&Q)
+
+	var invQz Fp2
+	invQz = Q.Z
+	inv(&invQz, &invQz)
+
+	mul(&pub3Pt[0], &P.X, &P.Z)
+	mul(&pub3Pt[1], &Q.X, &invQz)
+}
+
+
+// Construct Invalid public key tuple (P,Q) such that P and Q are in E[2^e2]
+// Simulate section 3.2 of paper https://eprint.iacr.org/2022/054.pdf
+// We only construct point P and Q because in the attacks the third point is P-Q by construction 
+// and the countermeasure does not test it
+// Without loss of generality, we assume the curve is the starting curve 
+func InvalidPKOrder2(scalar []uint8, pub3Pt *[3]Fp2) {
+	
+	var P, Q ProjectivePoint
+	
+	P = ProjectivePoint{X: params.A.AffineP, Z: params.OneFp2}
+	Q = ProjectivePoint{X: params.A.AffineQ, Z: params.OneFp2}
+
+	rand.Seed(time.Now().UnixNano())
+    random_index_p := rand.Intn(int(params.A.SecretByteLen-1)*8)
+	random_index_q := rand.Intn(int(params.A.SecretByteLen-1)*8)
+
+	P = montgomeryLadder(&params.InitCurve, &P, scalar, uint(random_index_p))
+	Q = montgomeryLadder(&params.InitCurve, &Q, scalar, uint(random_index_q))
+
+	var invQz, invPz Fp2
+	invQz = Q.Z
+	invPz = P.Z
+	inv(&invQz, &invQz)
+	inv(&invPz, &invPz)
+
+	mul(&pub3Pt[0], &P.X, &invPz)
+	mul(&pub3Pt[1], &Q.X, &invQz)
+}
+
+/* -------------------------------------------------------------------------
+   Public key / Ciphertext validation against attacks proposed in paper https://eprint.iacr.org/2022/054.pdf and HB
+   -------------------------------------------------------------------------*/
+
+func TestInvalidPK(t *testing.T) {
```

Comment:
here we can split this large test in three single (sub-)tests.

Reply:
sure, sounds good